src/js/controller/captions.js

```diff
@@ -42,36 +39,41 @@ const Captions = function(_model) {
         _tracks = [];
         _tracksById = {};
         _unknownCount = 0;
-    }
-
-    function _itemReadyHandler(item) {
-        // Clean up in case we're replaying
-        _itemHandler(_model, item);
 
+        const item = _model.get('playlistItem');
         const tracks = item.tracks;
         const len = tracks && tracks.length;
 
-        // Sideload tracks when not rendering natively
-        if (!_model.get('renderCaptionsNatively') && len) {
-            for (let i = 0; i < len; i++) {
-                /* eslint-disable no-loop-func */
-                const track = tracks[i];
-                if (_kindSupported(track.kind) && !_tracksById[track._id]) {
-                    _addTrack(track);
-                    loadFile(track,
-                        (vttCues) => {
-                            _addVTTCuesToTrack(track, vttCues);
-                        },
-                        (error) => {
-                            this.trigger(ERROR, {
-                                message: 'Captions failed to load',
-                                reason: error
-                            });
-                        });
+        // Update tracks once we know "renderCaptionsNatively" based on provider
+        _model.off('itemReady', null, this)
+            .once('itemReady', () => {
+                // Sideload tracks when not rendering natively
+                if (len && !_model.get('renderCaptionsNatively')) {
+                    for (let i = 0; i < len; i++) {
+                        /* eslint-disable no-loop-func */
+                        const track = tracks[i];
+                        if (_kindSupported(track.kind) && !_tracksById[track._id]) {
+                            _addTrack(track);
+                            loadFile(track,
+                                (vttCues) => {
+                                    _addVTTCuesToTrack(track, vttCues);
+                                },
+                                (error) => {
+                                    this.trigger(ERROR, {
+                                        message: 'Captions failed to load',
+                                        reason: error
+                                    });
+                                });
+                        }
+                    }
+                    _updateMenu();
```

Comment:
@egreaves I moved `_updateMenu` it outside the for loop. Any reason this native rendering `_addTrack` code can't be added to the provider's tracks mixin? It would be great to get rid of as many "itemReady" listeners as possible.

Reply:
Yeah, that should work since the menu would be setup [here](https://github.com/jwplayer/jwplayer/blob/f88b2dd2564e54b189a6ab310959c8b16980fd23/src/js/controller/controller.js#L687) after they get added by the mixin. Only risk I can think of is duplication, but I'm pretty sure there's logic to safeguard against that.